akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,493 @@
+ *
+ * The `ProducerController` sends the first message to the `ConsumerController` without waiting for
+ * a `Request` from the `ConsumerController`. The main reason for this is that when used with
+ * Cluster Sharding the first message will typically create the `ConsumerController`. It's
+ * also a way to connect the ProducerController and ConsumerController in a dynamic way, for
+ * example when the ProducerController is replaced.
+ *
+ * When the first message is received by the `ConsumerController` it sends back the initial `Request`,
+ * with demand of how many messages it can accept.
+ *
+ * Apart from the first message the `ProducerController` will not send more messages than requested
+ * by the `ConsumerController`.
+ *
+ * When there is demand from the consumer side the `ProducerController` sends `RequestNext` to the
+ * actual producer, which is then allowed to send one more message.
+ *
+ * Each message is wrapped by the `ProducerController` in [[ConsumerController.SequencedMessage]] with
+ * a monotonically increasing sequence number without gaps, starting at 1.
+ *
+ * In other words, the "request" protocol to the application producer and consumer is one-by-one, but
+ * between the `ProducerController` and `ConsumerController` it's window of messages in flight.
+ *
+ * The `Request` message also contains a `confirmedSeqNr` that is the acknowledgement
+ * from the consumer that it has received and processed all messages up to that sequence number.
+ *
+ * The `ConsumerController` will send [[ProducerController.Internal.Resend]] if a lost message is detected
+ * and then the `ProducerController` will resend all messages from that sequence number. The producer keeps
+ * unconfirmed messages in a buffer to be able to resend them. The buffer size is limited
+ * by the request window size.
+ *
+ * The resending is optional, and the `ConsumerController` can be started with `resendLost=false`
+ * to ignore lost messages, and then the `ProducerController` will not buffer unconfirmed messages.
+ * In that mode it provides only flow control but no reliable delivery.
+ */
+object ProducerController {
+
+  sealed trait InternalCommand
+
+  sealed trait Command[A] extends InternalCommand
+
+  final case class Start[A](producer: ActorRef[RequestNext[A]]) extends Command[A]
+
+  final case class RequestNext[A](
+      producerId: String,
+      currentSeqNr: Long,
+      confirmedSeqNr: Long,
+      sendNextTo: ActorRef[A],
+      askNextTo: ActorRef[MessageWithConfirmation[A]])
+
+  final case class RegisterConsumer[A](consumerController: ActorRef[ConsumerController.Command[A]]) extends Command[A]
+
+  /**
+   * For sending confirmation message back to the producer when the message has been fully delivered, processed,
+   * and confirmed by the consumer. Typically used with `ask` from the producer.
+   */
+  final case class MessageWithConfirmation[A](message: A, replyTo: ActorRef[Long]) extends InternalCommand
```

Comment:
`ActorRef[Long]` doesn't work for Java, I think. We can add some real class for that reply. Good anyway.